plugin/core/workspace.py

```diff
@@ -1,42 +1,116 @@
+    if project_file_name:
+        project_base_path = absolute_project_base_path(window)
+        path = ensure_absolute_path(project_base_path, path)
+    if not os.path.isdir(path):
+        raise ValueError("{} is not a directory".format(path))
+    return Workspace.from_path(path)
+
+
+def maybe_get_first_workspace_from_window(window: WindowLike) -> 'Optional[Workspace]':
+    try:
+        return get_first_workspace_from_window(window)
+    except Exception:
+        return None
+
+
+def maybe_get_workspace_from_view(view: ViewLike) -> 'Optional[Workspace]':
+    filename = view.file_name()
+    if filename and os.path.exists(filename):  # https://github.com/tomv564/LSP/issues/644
+        path = os.path.dirname(filename)
+        return Workspace.from_path(path)
+    debug("the current file isn't saved to disk.")
     return None
 
 
-def get_active_view_path(window: 'Any') -> 'Optional[str]':
+def get_workspaces_from_project_data(window: WindowLike) -> 'Optional[List[Workspace]]':
+    data = get_project_data_or_throw(window)
+    project_base_path = absolute_project_base_path(window)
+    folders = data.get("folders")
+    if folders is None:
+        raise ValueError("window folders is None")
+    return [workspace_from_sublime_project_data(project_base_path, folder) for folder in folders]
+
+
+def get_workspaces_from_window(window: WindowLike) -> 'Optional[List[Workspace]]':
+    project_file_name = window.project_file_name()
+    if project_file_name:
+        return get_workspaces_from_project_data(window)
+    try:
+        return [get_first_workspace_from_window(window)]
+    except AttributeError as error:
+        debug(error)
+    return None
+
+
+def get_common_prefix_of_workspaces(workspaces: 'Optional[List[Workspace]]') -> 'Optional[str]':
+    if workspaces is None:
+        return None
+    return get_common_parent(uri_to_filename(workspace.uri) for workspace in workspaces)
+
+
+def get_common_prefix_of_workspaces_for_window(window: WindowLike) -> 'Optional[str]':
+    return get_common_prefix_of_workspaces(get_workspaces_from_window(window))
+
+
+def get_common_parent(paths: 'Iterable[str]') -> str:
     """
     Returns the path containing the active view, if any.
     """
-    debug("couldn't determine project directory since no folders are open!")
-    view = window.active_view()
-    if view:
-        filename = view.file_name()
-        if filename and os.path.exists(filename):  # https://github.com/tomv564/LSP/issues/644
-            project_path = os.path.dirname(filename)
-            debug("using", project_path, "as a fallback.")
-            return project_path
-        else:
-            debug("no fallback path possible because the current file isn't saved to disk.")
-            return None
-    else:
-        debug("no view is active in current window")
-        return None  # https://github.com/tomv564/LSP/issues/219
+    return os.path.commonprefix([path + '/' for path in paths]).rstrip('/')
```

Comment:
`os.path.commonprefix` can return invalid paths (it does a character-based match instead of stopping at the last common directory).